bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/operations/OperationHistoryActionHandler.java

```diff
@@ -298,7 +303,32 @@ public final void run() {
 			if (getOperation() instanceof IAdvancedUndoableOperation2) {
 				runInBackground = ((IAdvancedUndoableOperation2) getOperation()).runInBackground();
 			}
-			progressDialog.run(runInBackground, true, runnable);
+			if (runInBackground) {
+				progressDialog.run(runInBackground, true, runnable);
+			} else {
+				final IJobManager manager = Job.getJobManager();
+				final ISchedulingRule rule = (getOperation() instanceof ISchedulableOperation)
+						? ((ISchedulableOperation) getOperation()).getSchedulingRule()
+						: null;
+				if (rule == null) {
+					progressDialog.run(runInBackground, true, runnable);
+				} else {
+					// prevent UI freeze during AutoBuild as in
+					// org.eclipse.jdt.internal.ui.refactoring.RefactoringExecutionHelper#perform
+					try {
+						try {
+							// interrupt autobuild or show Wait/Cancel Dialog:
+							Runnable r = () -> manager.beginRule(rule, null);
+							BusyIndicator.showWhile(parent.getDisplay(), r);
+						} catch (OperationCanceledException e) {
+							throw new InterruptedException(e.getMessage());
+						}
+						progressDialog.run(runInBackground, true, runnable); // <-- actual work
```

Comment:
From the javadoc:

> It is recommended that <code>fork</code> is set **to true in
> most cases.** If <code>fork</code> is set to <code>false</code>, the
> runnable will run in the UI thread and it is the runnable's
> responsibility to call <code>Display.readAndDispatch()</code> to ensure
> UI responsiveness.

so does the passed runable drives the event queue if runInBackround = false is passed in all cases?

Reply:
The 
`if (runInBackground) {..} else`
ensures runInBackground==false here

Reply:
if `runInBackground` is always `true` here ou should use a constant to prevent confusion.

Reply:
It's far better reading to pass (constant) variables then to call something(false, true,...) because it's not obvious without reading the called method what those values are used for.

Reply:
You can enable code-minings if you want.

Reply:
i don't want. i tried it. they are slow and move my code while i am reading :-(
i do like the idea though. sad it's not working good.